Handle props changes

ForbesLindesay opened the pull request:
[closes #128]

The peer dependency on final-form will need to be updated to the version that includes the `setConfig` method.

ForbesLindesay commented:
@erikras I assume you're going to require tests for this. Are you otherwise happy with the approach?

erikras commented:
I guess so, sure.

erikras commented:
Is this going anywhere? What I dislike about this PR is how "unminimizable" the code is. I question if the utility / bundle size ratio is high enough.

ForbesLindesay commented:
Sorry, I've been really busy. I'd be happy for someone to finish it off. I view this functionality as pretty critical as I've had multiple bugs caused by final-form not handling this case properly. It doesn't seem like much code to me.

Andarist commented:
could me shortened with smth like:
```js
updateConfig = (prev, current) => ['debug', 'mutators', 'onSubmit', 'validate', 'validateOnBlur'].forEach(key => {
	if (prev[key] === current[key]) {
		return
	}

	this.form.setConfig(key, current[key])
})
```

Impact of both versions on the gzipped production bundle would have to be measured though, gzip is a tricky thing.

From what I understand this would require bumping peerDep which would be a breaking change of sorts, so maybe some alternative approach could be used to solve this?

ForbesLindesay commented:
Changes:

1. updated based on @Andarist's suggestion to use an array for the keys.
2. ignore updates when `setConfig` does not exist, allowing the peer dependency to remain as it is and for this to be a non-breaking change.
3. added a test case. I've only tested `onSubmit` changes since the others are pretty much the same code path.

@erikras let me know if anything else is needed to get this landed, however I'm on holiday for a while after the end of this week, so if it needs substantial changes it may take a while for me to get to them.

erikras commented:
Looks great to me!

erikras commented:
Actually, there was a bug in your test. You were in initializing `state` to `newOnSubmit` instead of `oldOnSubmit`. 😄

Andarist commented:
@erikras it would be good to test gzipped size with previous version of this code

erikras commented:
@Andarist Nice! I originally had [the subscription filtering stuff](https://github.com/final-form/final-form/blob/master/src/subscriptionFilter.js) written out as 20 `if` blocks, because I couldn't figure out how to keep Flow happy, but when I moved it to iterating over an array of keys, I remember it dropping the bundle size substantially.

erikras commented:
Published in [`v3.2.1`](https://github.com/final-form/react-final-form/releases/tag/v3.2.1).
